plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -685,6 +685,18 @@ compileExpr e = traceCompilation 2 ("Compiling expr:" GHC.<+> GHC.ppr e) $ do
     _                  -> throwPlain $ CompilationError "No info for ByteString builtin"
 
   case e of
+    {- Note [Lazy boolean operators]
+      (||) and (&&) have a special treatment: we want them lazy in the second argument,
+      as this is the behavior in Haskell and other PLs.
+      Covered by this spec: plutus-tx-plugin/test/ShortCircuit/Spec.hs
+    -}
+    -- Lazy ||
+    GHC.App (GHC.App (GHC.Var fid) a) b | GHC.getOccString fid == "||" -> do
```

Comment:
- Rather than adding two cases here, you can use a similar approach to how we deal with things in `PlutusTx.Builtins.Internal`. I don't know if the latter is necessarily better, but it's better to be consistent, unless there's a strong reason not to.

- Don't use string literals here. They are error prone, and you aren't even checking the module names, which means that "||" matches any `||`, not just `PlutusTx.Bool.||`. It's much better to use TH name here - `'PlutusTx.Bool.||`.

- We need a Note explaining why we don't simply make use of `PlutusIR.Transform.RewriteRules`, and add a rewrite rule. That seems a simpler solution.

Reply:
> Rather than adding two cases here, you can use a similar approach to how we deal with things in PlutusTx.Builtins.Internal. I don't know if the latter is necessarily better, but it's better to be consistent, unless there's a strong reason not to.

I think these are rather different to how the builtins work, so it's fine for them to be here. We do have various special cases for Haskell functions in this area, but they're more special stuff like `unpackCString` or whatever.

> We need a Note explaining why we don't simply make use of PlutusIR.Transform.RewriteRules, and add a rewrite rule. That seems a simpler solution.

I think that might be worse, since we _can't_ match on the TH name there!

Reply:
Have a look at how the builtins do it. We already have a way of carrying around TH names

Reply:
> I think these are rather different to how the builtins work, so it's fine for them to be here.

I think they are exactly the same? When we see `PlutusTx.Builtins.Internal.addInteger`, we don't look at the unfolding, but directly interpret it as PIR term `(builtin addInteger)`. Here we directly interpret `||` as `if ... then ... else ...`.

> I think that might be worse, since we can't match on the TH name there!

Maybe but there still should be a Note

Reply:
Ok, I guess one difference than builtins is that we are rewriting the saturated applications of `||`, not `||` itself.

Reply:
Yes, this also works:
```haskell
  boolOperatorOr <- GHC.getName <$> getThing '(PlutusTx.Bool.||)
  boolOperatorAnd <- GHC.getName <$> getThing '(PlutusTx.Bool.&&)
  case e of
    {- Note [Lazy boolean operators]
      (||) and (&&) have a special treatment: we want them lazy in the second argument,
      as this is the behavior in Haskell and other PLs.
      Covered by this spec: plutus-tx-plugin/test/ShortCircuit/Spec.hs
    -}
    -- Lazy ||
    GHC.App (GHC.App (GHC.Var var) a) b | GHC.getName var == boolOperatorOr ->
      compileExpr $ GHC.mkIfThenElse a (GHC.Var GHC.trueDataConId) b
    -- Lazy &&
    GHC.App (GHC.App (GHC.Var var) a) b | GHC.getName var == boolOperatorAnd ->
      compileExpr $ GHC.mkIfThenElse a b (GHC.Var GHC.falseDataConId)
```